File: platforms/android/sdk/src/main/kotlin/dev/bota/sdk/WiFiManager.kt
package dev.bota.sdk

import dev.bota.sdk.internal.DeviceRuntime
import dev.bota.sdk.internal.bluetooth.BotaBluetoothUUIDs
import dev.bota.sdk.model.ConnectedDevice
import dev.bota.sdk.model.DeviceWiFiScanResult
import dev.bota.sdk.model.WiFiConfigResult
import dev.bota.sdk.model.WiFiScanUpdate
import dev.bota.sdk.model.WiFiStatusInfo
import java.util.UUID
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.CoroutineStart
import kotlinx.coroutines.Deferred
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.NonCancellable
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.TimeoutCancellationException
import kotlinx.coroutines.async
import kotlinx.coroutines.cancel
import kotlinx.coroutines.cancelAndJoin
import kotlinx.coroutines.channels.SendChannel
import kotlinx.coroutines.channels.awaitClose
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.callbackFlow
import kotlinx.coroutines.flow.collect
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.flow.mapNotNull
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import kotlinx.coroutines.withTimeout

public class WiFiManager internal constructor() {
    private data class ActiveOperation(
        val id: UUID,
        val task: Deferred<*>,
        val runtime: DeviceRuntime,
    )

    private data class StatusObserver(
        val runtime: DeviceRuntime,
        val deviceId: String,
        val channel: SendChannel<WiFiStatusInfo>,
        val task: Job,
    )

    private val lock = Any()
    private var runtime: DeviceRuntime? = null
    private var callbackScope: CoroutineScope? = null
    private var activeOperation: ActiveOperation? = null
    private val statusObservers = mutableMapOf<UUID, StatusObserver>()

    internal fun attach(runtime: DeviceRuntime) {
        synchronized(lock) {
            this.runtime = runtime
            callbackScope = CoroutineScope(SupervisorJob() + Dispatchers.Default)
        }
    }

    internal suspend fun detach() {
        val operation = synchronized(lock) { activeOperation }
        operation?.task?.cancelAndJoin()
        operation?.runtime?.operations?.end(operation.id)
        synchronized(lock) { if (activeOperation?.id == operation?.id) activeOperation = null }
        stopAllStatusObservers()
        synchronized(lock) {
            callbackScope?.cancel()
            callbackScope = null
            runtime = null
        }
    }

    public suspend fun configure(
        device: ConnectedDevice,
        ssid: String,
        password: String,
        grantBlob: String,
    ): WiFiConfigResult = performOperation(device, BotaOperation.Provision) { configured ->
        val grant = configured.createWiFiGrantPacket(grantBlob)
        val credentials = configured.createWiFiCredentialPacket(ssid, password)
        configured.directWrite(
            device.id,
            BotaBluetoothUUIDs.WifiService,
            BotaBluetoothUUIDs.WifiGrant,
            grant,
        )
        withSubscription(configured, device.id, BotaBluetoothUUIDs.WifiStatus) { notifications ->
            configured.directWrite(
                device.id,
                BotaBluetoothUUIDs.WifiService,
                BotaBluetoothUUIDs.WifiCredential,
                credentials,
            )
            awaitConfigResult(notifications, configured)
        }
    }

    public suspend fun disconnect(device: ConnectedDevice): WiFiConfigResult =
        performOperation(device, BotaOperation.Provision) { configured ->
            val command = configured.createWiFiCredentialPacket("", "")
            withSubscription(configured, device.id, BotaBluetoothUUIDs.WifiStatus) { notifications ->
                configured.directWrite(
                    device.id,
                    BotaBluetoothUUIDs.WifiService,
                    BotaBluetoothUUIDs.WifiCredential,
                    command,
                )
                awaitConfigResult(notifications, configured)
            }
        }

    public suspend fun readStatus(device: ConnectedDevice): WiFiStatusInfo =
        performOperation(device, BotaOperation.ReadStatus) { configured ->
            configured.parseWiFiStatusInfo(
                configured.directRead(
                    device.id,
                    BotaBluetoothUUIDs.WifiService,
                    BotaBluetoothUUIDs.WifiStatus,
                ),
            )
        }

    public fun statusUpdates(device: ConnectedDevice): Flow<WiFiStatusInfo> {
        val configured = configuredRuntime()
        configured.connection.require(device)
        configured.authorize(BotaOperation.ReadStatus)
        val cleanupScope = synchronized(lock) { callbackScope } ?: unavailable()
        return callbackFlow {
            val source = configured.directSubscribe(
                device.id,
                BotaBluetoothUUIDs.WifiService,
                BotaBluetoothUUIDs.WifiStatus,
            )
            val id = UUID.randomUUID()
            val collector = launch(start = CoroutineStart.LAZY) {
                try {
                    source.collect { send(configured.parseWiFiStatusInfo(it)) }
                    close()
                } catch (error: CancellationException) {
                    close()
                    throw error
                } catch (error: Throwable) {
                    close(error)
                } finally {
                    withContext(NonCancellable) {
                        stopStatusObserver(id, cancelTask = false, closeChannel = false)
                    }
                }
            }
            synchronized(lock) {
                statusObservers[id] = StatusObserver(configured, device.id, channel, collector)
            }
            collector.start()
            awaitClose {
                collector.cancel()
                cleanupScope.launch { stopStatusObserver(id, cancelTask = true, closeChannel = true) }
            }
        }
    }

    public suspend fun scanNetworks(device: ConnectedDevice): DeviceWiFiScanResult =
        performOperation(device, BotaOperation.ReadStatus) { configured ->
            val command = configured.createWiFiScanCommand()
            withSubscription(configured, device.id, BotaBluetoothUUIDs.WifiScan) { notifications ->
                configured.directWrite(
                    device.id,
                    BotaBluetoothUUIDs.WifiService,
                    BotaBluetoothUUIDs.WifiScan,
                    command,
                )
                awaitScanResult(notifications, configured)
            }
        }

    public suspend fun cancelCurrentOperation() {
        val operation = synchronized(lock) { activeOperation } ?: return
        operation.task.cancelAndJoin()
        operation.runtime.operations.end(operation.id)
        synchronized(lock) { if (activeOperation?.id == operation.id) activeOperation = null }
    }

    private suspend fun <T> performOperation(
        device: ConnectedDevice,
        operation: BotaOperation,
        body: suspend (DeviceRuntime) -> T,
    ): T {
        val configured = configuredRuntime()
        configured.connection.require(device)
        configured.authorize(operation)
        val id = UUID.randomUUID()
        configured.operations.begin(id, operation)
        return kotlinx.coroutines.coroutineScope {
            val task = async(start = CoroutineStart.LAZY) { body(configured) }
            synchronized(lock) { activeOperation = ActiveOperation(id, task, configured) }
            task.start()
            try {
                task.await()
            } catch (error: CancellationException) {
                throw cancelled(operation)
            } finally {
                synchronized(lock) { if (activeOperation?.id == id) activeOperation = null }
                configured.operations.end(id)
            }
        }
    }

    private suspend fun <T> withSubscription(
        configured: DeviceRuntime,
        deviceId: String,
        characteristic: UUID,
        body: suspend (Flow<ByteArray>) -> T,
    ): T {
        val source = configured.directSubscribe(deviceId, BotaBluetoothUUIDs.WifiService, characteristic)
        return try {
            body(source)
        } finally {
            withContext(NonCancellable) {
                runCatching {
                    configured.directUnsubscribe(deviceId, BotaBluetoothUUIDs.WifiService, characteristic)
                }
            }
        }
    }

    private suspend fun awaitConfigResult(
        notifications: Flow<ByteArray>,
        configured: DeviceRuntime,
    ): WiFiConfigResult = try {
        withTimeout(operationTimeoutMilliseconds) {
            notifications.mapNotNull { runCatching { configured.parseWiFiConfigResult(it) }.getOrNull() }.first()
        }
    } catch (_: TimeoutCancellationException) {
        throw timeout("WiFi configuration")
    } catch (_: NoSuchElementException) {
        throw endedWithoutResult("WiFi configuration")
    }

    private suspend fun awaitScanResult(
        notifications: Flow<ByteArray>,
        configured: DeviceRuntime,
    ): DeviceWiFiScanResult = try {
        withTimeout(operationTimeoutMilliseconds) {
            notifications.mapNotNull {
                when (val update = configured.parseWiFiScanResult(it)) {
                    is WiFiScanUpdate.Pending -> null
                    is WiFiScanUpdate.Done -> update.result
                }
            }.first()
        }
    } catch (_: TimeoutCancellationException) {
        throw timeout("WiFi scan")
    } catch (_: NoSuchElementException) {
        throw endedWithoutResult("WiFi scan")
    }

    private suspend fun stopStatusObserver(id: UUID, cancelTask: Boolean, closeChannel: Boolean) {
        val observer = synchronized(lock) { statusObservers.remove(id) } ?: return
        if (cancelTask) observer.task.cancel()
        if (closeChannel) observer.channel.close()
        runCatching {
            observer.runtime.directUnsubscribe(
                observer.deviceId,
                BotaBluetoothUUIDs.WifiService,
                BotaBluetoothUUIDs.WifiStatus,
            )
        }
    }

    private suspend fun stopAllStatusObservers() {
        val observers = synchronized(lock) { statusObservers.values.toList().also { statusObservers.clear() } }
        observers.forEach {
            it.task.cancel()
            it.channel.close()
            runCatching {
                it.runtime.directUnsubscribe(
                    it.deviceId,
                    BotaBluetoothUUIDs.WifiService,
                    BotaBluetoothUUIDs.WifiStatus,
                )
            }
        }
    }

    private fun configuredRuntime(): DeviceRuntime = synchronized(lock) { runtime } ?: unavailable()

    private companion object {
        fun timeout(label: String): BotaSDKError.Core = BotaSDKError.Core(
            BotaErrorCode.Timeout,
            BotaOperation.Provision,
            retryable = true,
            protocolStatus = null,
            detail = "$label timed out",
        )

        fun endedWithoutResult(label: String): BotaSDKError.Core = BotaSDKError.Core(
            BotaErrorCode.UnexpectedEvent,
            BotaOperation.Provision,
            retryable = true,
            protocolStatus = null,
            detail = "$label ended without a result",
        )

        fun cancelled(operation: BotaOperation): BotaSDKError.Core = BotaSDKError.Core(
            BotaErrorCode.Cancelled,
            operation,
            retryable = true,
            protocolStatus = null,
            detail = "WiFi operation was cancelled",
        )
    }
}

private const val operationTimeoutMilliseconds: Long = 30_000

File: platforms/apple/Sources/BotaAppleSDK/WiFiManager.swift
import Foundation

public actor WiFiManager {
    private struct ActiveOperation {
        let id: UUID
        let cancel: @Sendable () -> Void
        let wait: @Sendable () async -> Void
    }

    private struct StatusObserver {
        let continuation: AsyncThrowingStream<WiFiStatusInfo, Error>.Continuation
        let lease: WiFiSubscriptionLease
        var task: Task<Void, Never>?
    }

    private static let operationTimeoutNanoseconds: UInt64 = 30_000_000_000

    private var runtime: DeviceRuntime?
    private var activeOperation: ActiveOperation?
    private var statusObservers: [UUID: StatusObserver] = [:]

    public init() {}

    func attach(_ runtime: DeviceRuntime) { self.runtime = runtime }

    func detach() async {
        let configured = runtime
        if let operation = activeOperation {
            operation.cancel()
            await operation.wait()
            await configured?.operations.end(operation.id)
            activeOperation = nil
        }
        await stopAllStatusObservers()
        runtime = nil
    }

    public func configure(
        _ device: ConnectedDevice,
        ssid: String,
        password: String,
        grantBlob: String
    ) async throws -> WiFiConfigResult {
        try await performOperation(device, operation: .provision) { runtime in
            let grant = try runtime.createWiFiGrantPacket(grantBlob)
            let credentials = try runtime.createWiFiCredentialPacket(ssid, password)
            try await runtime.directWrite(
                device.id,
                BotaBluetoothUUIDs.wifiService,
                BotaBluetoothUUIDs.wifiGrant,
                grant
            )
            return try await Self.withSubscription(
                runtime: runtime,
                deviceID: device.id,
                characteristicUUID: BotaBluetoothUUIDs.wifiStatus
            ) { notifications in
                try await runtime.directWrite(
                    device.id,
                    BotaBluetoothUUIDs.wifiService,
                    BotaBluetoothUUIDs.wifiCredential,
                    credentials
                )
                return try await Self.awaitConfigResult(notifications, runtime: runtime)
            }
        }
    }

    public func disconnect(_ device: ConnectedDevice) async throws -> WiFiConfigResult {
        try await performOperation(device, operation: .provision) { runtime in
            let command = try runtime.createWiFiCredentialPacket("", "")
            return try await Self.withSubscription(
                runtime: runtime,
                deviceID: device.id,
                characteristicUUID: BotaBluetoothUUIDs.wifiStatus
            ) { notifications in
                try await runtime.directWrite(
                    device.id,
                    BotaBluetoothUUIDs.wifiService,
                    BotaBluetoothUUIDs.wifiCredential,
                    command
                )
                return try await Self.awaitConfigResult(notifications, runtime: runtime)
            }
        }
    }

    public func readStatus(_ device: ConnectedDevice) async throws -> WiFiStatusInfo {
        try await performOperation(device, operation: .readStatus) { runtime in
            let data = try await runtime.directRead(
                device.id,
                BotaBluetoothUUIDs.wifiService,
                BotaBluetoothUUIDs.wifiStatus
            )
            return try runtime.parseWiFiStatusInfo(data)
        }
    }

    public func statusUpdates(
        _ device: ConnectedDevice
    ) async throws -> AsyncThrowingStream<WiFiStatusInfo, Error> {
        let runtime = try configuredRuntime()
        try await runtime.connection.require(device)
        let source = try await runtime.directSubscribe(
            device.id,
            BotaBluetoothUUIDs.wifiService,
            BotaBluetoothUUIDs.wifiStatus
        )
        let lease = WiFiSubscriptionLease(
            runtime: runtime,
            deviceID: device.id,
            characteristicUUID: BotaBluetoothUUIDs.wifiStatus
        )
        let id = UUID()
        let pair = AsyncThrowingStream<WiFiStatusInfo, Error>.makeStream()
        statusObservers[id] = StatusObserver(continuation: pair.continuation, lease: lease, task: nil)
        let task = Task {
            do {
                for try await data in source {
                    pair.continuation.yield(try runtime.parseWiFiStatusInfo(data))
                }
                pair.continuation.finish()
            } catch is CancellationError {
                pair.continuation.finish()
            } catch {
                pair.continuation.finish(throwing: error)
            }
            await self.completeStatusObserver(id)
        }
        statusObservers[id]?.task = task
        pair.continuation.onTermination = { @Sendable _ in
            Task { await self.stopStatusObserver(id) }
        }
        return pair.stream
    }

    public func scanNetworks(_ device: ConnectedDevice) async throws -> DeviceWiFiScanResult {
        try await performOperation(device, operation: .readStatus) { runtime in
            let command = try runtime.createWiFiScanCommand()
            return try await Self.withSubscription(
                runtime: runtime,
                deviceID: device.id,
                characteristicUUID: BotaBluetoothUUIDs.wifiScan
            ) { notifications in
                try await runtime.directWrite(
                    device.id,
                    BotaBluetoothUUIDs.wifiService,
                    BotaBluetoothUUIDs.wifiScan,
                    command
                )
                return try await Self.awaitScanResult(notifications, runtime: runtime)
            }
        }
    }

    public func cancelCurrentOperation() async {
        guard let operation = activeOperation else { return }
        operation.cancel()
        await operation.wait()
        await finishOperation(operation.id, runtime: runtime)
    }

    private func performOperation<T: Sendable>(
        _ device: ConnectedDevice,
        operation: BotaOperation,
        body: @escaping @Sendable (DeviceRuntime) async throws -> T
    ) async throws -> T {
        let runtime = try configuredRuntime()
        try await runtime.connection.require(device)
        let id = UUID()
        try await runtime.operations.begin(id, operation: operation)
        let task = Task { try await body(runtime) }
        activeOperation = ActiveOperation(
            id: id,
            cancel: { task.cancel() },
            wait: { _ = try? await task.value }
        )
        do {
            let value = try await withTaskCancellationHandler {
                try await task.value
            } onCancel: {
                task.cancel()
            }
            await finishOperation(id, runtime: runtime)
            return value
        } catch is CancellationError {
            await finishOperation(id, runtime: runtime)
            throw Self.cancelled(operation)
        } catch {
            await finishOperation(id, runtime: runtime)
            throw error
        }
    }

    private func finishOperation(_ id: UUID, runtime: DeviceRuntime?) async {
        if activeOperation?.id == id { activeOperation = nil }
        await runtime?.operations.end(id)
    }

    private func configuredRuntime() throws -> DeviceRuntime {
        guard let runtime else {
            throw BotaSDKError(
                code: .featureUnavailable,
                operation: .validate,
                retryable: false,
                detail: "BotaDeviceClient.configure() must be called first"
            )
        }
        return runtime
    }

    private func completeStatusObserver(_ id: UUID) async {
        guard let observer = statusObservers.removeValue(forKey: id) else { return }
        await observer.lease.close()
    }

    private func stopStatusObserver(_ id: UUID) async {
        guard let observer = statusObservers.removeValue(forKey: id) else { return }
        observer.task?.cancel()
        observer.continuation.finish()
        await observer.lease.close()
    }

    private func stopAllStatusObservers() async {
        let observers = statusObservers
        statusObservers.removeAll()
        for observer in observers.values {
            observer.task?.cancel()
            observer.continuation.finish()
            await observer.lease.close()
        }
    }

    private static func withSubscription<T: Sendable>(
        runtime: DeviceRuntime,
        deviceID: String,
        characteristicUUID: String,
        body: @escaping @Sendable (AsyncThrowingStream<Data, Error>) async throws -> T
    ) async throws -> T {
        let source = try await runtime.directSubscribe(
            deviceID,
            BotaBluetoothUUIDs.wifiService,
            characteristicUUID
        )
        let lease = WiFiSubscriptionLease(
            runtime: runtime,
            deviceID: deviceID,
            characteristicUUID: characteristicUUID
        )
        return try await withTaskCancellationHandler {
            do {
                let value = try await body(source)
                await lease.close()
                return value
            } catch {
                await lease.close()
                throw error
            }
        } onCancel: {
            Task { await lease.close() }
        }
    }

    private static func awaitConfigResult(
        _ notifications: AsyncThrowingStream<Data, Error>,
        runtime: DeviceRuntime
    ) async throws -> WiFiConfigResult {
        try await withThrowingTaskGroup(of: WiFiConfigResult.self) { group in
            group.addTask {
                for try await data in notifications {
                    if let result = try? runtime.parseWiFiConfigResult(data) { return result }
                }
                throw endedWithoutResult("WiFi configuration")
            }
            group.addTask {
                try await Task.sleep(nanoseconds: operationTimeoutNanoseconds)
                throw timeout("WiFi configuration")
            }
            let result = try await group.next()!
            group.cancelAll()
            return result
        }
    }

    private static func awaitScanResult(
        _ notifications: AsyncThrowingStream<Data, Error>,
        runtime: DeviceRuntime
    ) async throws -> DeviceWiFiScanResult {
        try await withThrowingTaskGroup(of: DeviceWiFiScanResult.self) { group in
            group.addTask {
                for try await data in notifications {
                    switch try runtime.parseWiFiScanResult(data) {
                    case .pending:
                        continue
                    case let .done(result):
                        return result
                    }
                }
                throw endedWithoutResult("WiFi scan")
            }
            group.addTask {
                try await Task.sleep(nanoseconds: operationTimeoutNanoseconds)
                throw timeout("WiFi scan")
            }
            let result = try await group.next()!
            group.cancelAll()
            return result
        }
    }

    private static func timeout(_ label: String) -> BotaSDKError {
        BotaSDKError(
            code: .timeout,
            operation: .provision,
            retryable: true,
            detail: "\(label) timed out"
        )
    }

    private static func endedWithoutResult(_ label: String) -> BotaSDKError {
        BotaSDKError(
            code: .unexpectedEvent,
            operation: .provision,
            retryable: true,
            detail: "\(label) ended without a result"
        )
    }

    private static func cancelled(_ operation: BotaOperation) -> BotaSDKError {
        BotaSDKError(
            code: .cancelled,
            operation: operation,
            retryable: true,
            detail: "WiFi operation was cancelled"
        )
    }
}

private actor WiFiSubscriptionLease {
    private let runtime: DeviceRuntime
    private let deviceID: String
    private let characteristicUUID: String
    private var closed = false

    init(runtime: DeviceRuntime, deviceID: String, characteristicUUID: String) {
        self.runtime = runtime
        self.deviceID = deviceID
        self.characteristicUUID = characteristicUUID
    }

    func close() async {
        guard !closed else { return }
        closed = true
        try? await runtime.directUnsubscribe(
            deviceID,
            BotaBluetoothUUIDs.wifiService,
            characteristicUUID
        )
    }
}
